src/rgw/store/sfs/CMakeLists.txt

```diff
@@ -0,0 +1,53 @@
+#need to update cmake version here
+cmake_minimum_required(VERSION 3.14.0)
+project(sfsstore)
+
+option(USE_SFS "Enable SFS" ON)
+
+set (CMAKE_INCLUDE_DIR ${CMAKE_INCLUDE_DIR} "${CMAKE_CURRENT_SOURCE_DIR}/sqlite")
+
+set(sfsstore_sqlite_srcs
+    sqlite/sqlite_users.cc
+    sqlite/sqlite_buckets.cc
+    sqlite/sqlite_objects.cc
+    sqlite/sqlite_versioned_objects.cc
+    sqlite/users/users_conversions.cc
+    sqlite/buckets/bucket_conversions.cc
+    sqlite/objects/object_conversions.cc
+    sqlite/versioned_object/versioned_object_conversions.cc
+    )
+
+set(sfsstore_srcs
+    bucket.cc
+    multipart.cc
+    object.cc
+    user.cc
+    types.cc
+    zone.cc
+    writer.cc
+    sfs_bucket.cc
+    sfs_user.cc
+    )
+
+add_library(sfsstore_lib ${sfsstore_sqlite_srcs})
+target_include_directories(sfsstore_lib PUBLIC "${CMAKE_SOURCE_DIR}/src/fmt/include")
+target_include_directories(sfsstore_lib PUBLIC "${CMAKE_SOURCE_DIR}/src/rgw")
+set(link_targets spawn)
+if(WITH_JAEGER)
+  list(APPEND link_targets jaeger_base)
+endif()
+target_link_libraries(sfsstore_lib PUBLIC ${link_targets})
+
+set (CMAKE_LINK_LIBRARIES ${CMAKE_LINK_LIBRARIES} sfsstore_lib)
+
+IF(USE_SQLITE)
+  add_compile_definitions(SQLITE_ENABLED=1)
+  set (CMAKE_LINK_LIBRARIES ${CMAKE_LINK_LIBRARIES} rgw_common)
+ENDIF()
+
+# add pthread library
+set (CMAKE_LINK_LIBRARIES ${CMAKE_LINK_LIBRARIES} pthread)
+
+include_directories(${CMAKE_INCLUDE_DIR})
+add_library(sfsstore STATIC ${sfsstore_srcs})
```

Comment:
I think we should have the sqlite bits here as well, instead of having them as a separate library.

Reply:
Correct me if I'm wrong: now `sfs` be relies on `dbstore` for the build of `sqlite`, correct?
So, if I hev understand correctly we cannot build `sfs` if we are not building also `dbstore`; if this is true (and it should) I think this is wrong, we should make `sfs` independent.
I see that sqlite files are placed inside `store/dbstore/sqlite`; do you think should we make the same for `sfs`?

Reply:
@jecluis @0xavi0 
UPDATE

with the last push, you can now build `radosgw` with `dbstore` and `sfs` independently.
Before this, you were forced to specify: `-DWITH_RADOSGW_DBSTORE=ON` to be able to link `sfs` with `sqlite`.

In more detail, the patch consists in adding these directives to `ceph/src/rgw/store/sfs/CMakeLists.txt`:

```
find_package(SQLite3 REQUIRED)
set(SQLITE_COMPILE_FLAGS "-DSQLITE_THREADSAFE=1")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${SQLITE_COMPILE_FLAGS}")
```

Because this is included also in `ceph/src/rgw/store/dbstore/sqlite/CMakeLists.txt` I have a concern that this could potentially clash in the future if this should diverge in the two BEs.
Depending on what is built first, this could lead to linking with a wrong sqlite lib.
IMO, this is better btw, to blindly rely on a `sqlite` lib compiled in another backend.
Said this, because `sfs` and `dbstore` both rely on `sqlite` to work, and because we cannot guarantee that they will diverge in the future, I would say that, logically, `radosgw` should be mutually built with one or the other (but not both).

Reply:
I find this odd because, if we have a required package, I would expect this package to be installed in the system rather than having to compile it -- in which case the compile flags would be moot. Is this not the case?

Reply:
cmake is building sqlite from the source, I don't think it is checking if there is a sqlite system package installed on the host.
If you try building `sfs` with `-DWITH_RADOSGW_DBSTORE=OFF` your build will fail because it cannot link with sqlite.

Reply:
but isn't that because it's the dbstore cmake file that is defining the `find_package(SQLite3 REQUIRED)`, rather than the build stuff?

Reply:
This is interesting, I'm not a cmake expert so I don't know exactly what `find_package` means.
Because in dbstore cmake I found: 

```
find_package(SQLite3 REQUIRED)
set(SQLITE_COMPILE_FLAGS "-DSQLITE_THREADSAFE=1")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${SQLITE_COMPILE_FLAGS}")
```

I assumed that it was actually compiling it; but I could be wrong.
What I made is to move these directive on `sfs` cmake too.

Reply:
Mind trying compiling the sfs code (without dbstore) just with the `find_package(...)` bit and without the flags?

Reply:
Build has been successful.
I guess in this case `sqlite` has been built without support for multithreading.

Reply:
I'm still not getting where you are seeing sqlite being compiled. I can't find any reference to that in ceph, aside from depends on the rpm spec file (which indicates that we expect the package to be available, instead of being compiled).

Can you verify that there's a sqlite being built without multithreading?

Reply:
@jecluis @0xavi0 
I was wondering: the first time we developed the `sqlite` "layer" with the first mappings, did this question arise?
I mean did we just assume to have `sqlite` available _gratis_ from some other part of the `radosgw` ?
Did we assume the available `sqlite` was compiled with `-DSQLITE_THREADSAFE=1`?

Reply:
I did assume we'd have sqlite **requirement** for free from Ceph, not radosgw itself, because there are other portions of ceph that rely on sqlite. And I did assume we'd have the **linking** for free from rgw.

As for the threadsafe bit, I recall Tim checked it was being indeed compiled with that in OBS.

Reply:
I just have had a chat with @tserong and we dug this a bit.
This is what we found:

there is a file `FindSQLite3.cmake` that is providing a mapping between the name `SQLite3` and the system library name: `sqlite3`.
When in a `CMakeLists.txt` there is the directive:  `find_package(SQLite3 REQUIRED)`, cmake will link the sqlite target system library with the final executable.
As you can see [here](https://build.opensuse.org/public/build/openSUSE:Factory/standard/x86_64/sqlite3/_log), on our systems, the package is built with the threadsafe capability enabled.
What is more strange to all us is that there are these two lines on the dbstore `CMakeLists.txt`: 

```
set(SQLITE_COMPILE_FLAGS "-DSQLITE_THREADSAFE=1")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${SQLITE_COMPILE_FLAGS}")
```
This because, normally these flags are used when compiling the `sqlite` library.
Our guess is that this has been added to make sure these flags are used also in the linking phase.
I checked if in the `sqlite.h` there are `ifdef`s related to `SQLITE_THREADSAFE` flag, but it seems me not.
As final thought for this PR, I would be of the idea to leave in the `sfs` `CMakeLists.txt` the flags (even if potentially useless).

Reply:
added a comment regarding this in `src/rgw/store/sfs/CMakeLists.txt`